### PR TITLE
Implementing delete function for photo

### PR DESCRIPTION
**Pull request by stephensxu**
- Implement delete function for photo; deleting photo remove database record and file on aws S3.
- Implement function to assign front_page_photo to kitchen

### CODE REVIEW

File: app/assets/stylesheets/kitchens_index_page.scss [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #logged_in_nav_list {
   padding-right: 10px;
   font-weight: bold;
+  font-size: 15px;
```
Comment (jfarmer) [inline app/assets/stylesheets/kitchens_index_page.scss:15]: This change seems unrelated to the "photo delete" functionality.

File: app/controllers/photos_controller.rb [hunk 2]
```diff
@@ -15,6 +21,17 @@ def new
     @photo = Photo.new
   end
 
+  def edit
+  end
+
+  def destroy
```
Comment (jfarmer) [inline app/controllers/photos_controller.rb:27]: Be careful with this code.  You're assuming everything will work.  What happens if `@photo.remove_picture!` succeeds but `@photo.destroy` fails?  What happens if `@photo.remove_picture!` fails but `@photo.destroy` succeeds?

In either case, your application will be in an inconsistent state.

Also, per the documentation, calling `Photo#destroy` will automatically call `remove_picture!`.  See https://github.com/carrierwaveuploader/carrierwave#skipping-activerecord-callbacks.

Comment (jfarmer) [inline app/controllers/photos_controller.rb:27]: Lastly, did you check to make sure this code handles the case where a particular `Photo` is the front-page photo of a `Kitchen`?  Is the `front_page_photo_id` field updated accordingly on the `kitchens` table or does it now contain a dangling reference?

This is why model tests are valuable.  You'd say something like this in RSpec

``` ruby
expect {
  front_page_photo.destroy
}.to change { kitchen.front_page_photo }.to(nil)
```

Don't quote me on that verbatim.  You might need to reload the `kitchen` object from the database, for example.

Comment (jfarmer) [inline app/controllers/photos_controller.rb:27]: Remember: writing tests are valuable not just because they protect against errors, but because they force you to think about what you code is _actually_ doing.  Over the course of writing a test you'll often realize that you forgot to account for something in your code that didn't occur to you when you originally wrote the code.

File: app/models/kitchen.rb [hunk 3]
```diff
@@ -76,4 +76,15 @@ def archive!
...
     self.data_status == 'archive'
   end
+
+  def assign_front_page_photo(photo)
+    if self.photos.count == 1
+      self.update_attributes(:front_page_photo_id => photo.id)
+    end
+  end
+
+  def reassign_front_page_photo(photo)
```
Comment (jfarmer) [inline app/models/kitchen.rb:86]: Uh, you can just do

``` ruby
kitchen.front_page_photo = some_photo
```

File: app/models/kitchen.rb [hunk 4]
```diff
@@ -76,4 +76,15 @@ def archive!
   def archived?
     self.data_status == 'archive'
   end
+
+  def assign_front_page_photo(photo)
+    if self.photos.count == 1
+      self.update_attributes(:front_page_photo_id => photo.id)
```
Comment (jfarmer) [inline app/models/kitchen.rb:82]: This belongs in an ActiveRecord `after_create` callback.  Something like...

``` ruby
class Photo < ActiveRecord::Base
  after_create :set_as_front_page_photo_if_first

  private

  def set_as_front_page_photo_if_first
    kitchen.update!(front_page_photo: self) if kitchen.photos.count == 1
  end
end
```

`update!` with the bang `!` will raise an error if something fails whereas `update` without the bang will only return `false`.  This is almost certainly what you want, since something has gone terribly wrong if `kitchen.update!` fails.

File: app/models/kitchen.rb (hunk 3 above)
Comment (jfarmer) [inline app/models/kitchen.rb:86]: Note that

``` ruby
kitchen.front_page_photo = some_photo
```

_won't_ save the `kitchen` record to the database.  You'd have to still call `kitchen.save`.

Still, I'm imagining the controller code to handle this.  You should be issuing a `PUT /kitchens/:kitchen_id` request with, e.g., `params == { front_page_photo_id: "123" }`.  Assuming you have the correct validations on the `Kitchen` model, i.e., you can't assign a front-page photo to a `Kitchen` unless that photo is one of the `Kitchens` photos, then your `KitchensController#update` action shouldn't have to change at _all_ to account for `front_page_photo_id`.  As far as its concerned, it's just another "update" request that happens to contain the `front_page_photo_id` param.

File: app/controllers/photos_controller.rb (hunk 2 above)
Comment (stephensxu) [inline app/controllers/photos_controller.rb:27]: You are right; Photo#destroy will automatically call remove_picture! Also, it does not delete the reference if a particular photo is front_page_photo of the kitchen. As a solution I redefined the method like this(not sure if it's the best way):

```
  def destroy
    @photo = Photo.find(params[:id])
    @kitchen = @photo.kitchen
    @kitchen.front_page_photo.destroy if @photo == @kitchen.front_page_photo
    @photo.destroy
    redirect_to kitchen_photos_path(@kitchen)
  end
```

File: app/models/kitchen.rb (hunk 3 above)
Comment (stephensxu) [inline app/models/kitchen.rb:86]: A bug related to this issue cost me 3 hours last night to realize that `kitchen.ront_page_photo = some_photo` doesn't work by itself. Only that I read your comment after realizing that...sad.
